pkg/filesystem/subvolume.go

```diff
@@ -201,17 +208,100 @@ func unMarshaljson(list string) []fsStruct {
+		}
+	}
+	return "", fmt.Errorf("pool not found")
+}
+
+// deleteOmap deletes omap object and key for the given subvolume.
+func deleteOmap(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNamespace, CephClusterNamespace, subVol, fs string) {
+	logging.Info("Deleting the omap object and key for subvolume %q", subVol)
+	omapdetails := getOmapDetails(ctx, clientsets, OperatorNamespace, CephClusterNamespace, subVol, fs)
+	poolName, err := getPoolName(ctx, clientsets, OperatorNamespace, CephClusterNamespace, fs)
+	if err != nil || poolName == "" {
+		logging.Fatal(fmt.Errorf("pool name not found: %q", err))
+	}
+	if omapdetails.omapval != "" {
+		// remove omap object.
+		_, err := exec.RunCommandInOperatorPod(ctx, clientsets, "rados", []string{"rm", omapdetails.omapval, "-p", poolName, "--namespace", "csi"}, OperatorNamespace, CephClusterNamespace, true)
+		if err != nil {
+			logging.Error(err, "failed to remove omap object for subvolume %q", subVol)
+		}
+		logging.Info("omap object:%q deleted", omapdetails.omapval)
+	}
+	if omapdetails.omapkey != "" {
+		// remove omap key.
+		_, err := exec.RunCommandInOperatorPod(ctx, clientsets, "rados", []string{"rmomapkey", "csi.volumes.default", omapdetails.omapkey, "-p", poolName, "--namespace", "csi"}, OperatorNamespace, CephClusterNamespace, true)
+		if err != nil {
+			logging.Error(err, "failed to remove omap key for subvolume %q", subVol)
+		}
+		logging.Info("omap key:%q deleted", omapdetails.omapkey)
+	}
+}
+
+// getOmapDetails gets the omap key and value details for a given subvolume.
+// Deletion of omap object required the subvolumeName which is of format
+// csi.volume.subvolume, where subvolume is the name of subvolume that needs to be
+// deleted.
+// similarly to delete of omap key requires csi.volume.ompakey, where
+// omapkey is the pv name which is extracted the omap object.
+func getOmapDetails(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNamespace, CephClusterNamespace, subVol, fs string) omapDetails {
+
+	var details omapDetails
+	var omapval, omapkey string
+	poolName, err := getPoolName(ctx, clientsets, OperatorNamespace, CephClusterNamespace, fs)
+	if err != nil || poolName == "" {
+		logging.Fatal(fmt.Errorf("pool name not found: %q", err))
+	}
+	// omapval is of format csi.volume.427774b4-340b-11ed-8d66-0242ac110005
+	// which is similar to volume name csi-vol-427774b4-340b-11ed-8d66-0242ac110005
+	// hence, replacing '-' to '.' to get the omapval
+	replace := strings.Replace(subVol, "-", ".", 2)
+	omapval = strings.Replace(replace, "vol", "volume", 1)
+
+	cmd := []string{"getomapval", omapval, "csi.volname", "-p", poolName, "--namespace", "csi", "/dev/stdout"}
+	pvname, err := exec.RunCommandInOperatorPod(ctx, clientsets, "rados", cmd, OperatorNamespace, CephClusterNamespace, true)
+	if err != nil || pvname == "" {
+		if !strings.Contains(err.Error(), "No such file or directory") {
+			logging.Fatal(fmt.Errorf("failed to list omapvals: %q", err))
```

Comment:
You are not listing omap values anymore, you now only get a single one. Include the missing value and error?

Is this really fatal? What if the values were cleaned up some way already?